File: 1770.maximum-score-from-performing-multiplication-operations/solution.py
#
# TC: O(M * M)
# SC: O(M * M)
#
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        N = len(nums)
        M = len(multipliers)

        @cache
        def dp(mIdx: int, L: int) -> int:
            R = N - 1 - (mIdx - L)
            if (mIdx == M):
                return 0

            doLeft = nums[L] * multipliers[mIdx] + dp(mIdx + 1, L + 1)
            doRight = nums[R] * multipliers[mIdx] + dp(mIdx + 1, L)
            return max(doLeft, doRight)
        return dp(0, 0)

File: 1770.maximum-score-from-performing-multiplication-operations/solution.ts
function maximumScore(nums: number[], multipliers: number[]): number {

  const cache = new Map<string, number>();
  
  const backtrack = (left: number, right: number, idx: number): number => {
      const key = `${left}_${right}`;
      if (cache.has(key)) {
          return cache.get(key)!;
      }
      if (left > right || idx === multipliers.length) {
          return 0;
      }
      
      const leftResult = backtrack(left + 1, right, idx + 1) + (nums[left] * multipliers[idx]);
      const rightResult = backtrack(left, right - 1, idx + 1) + (nums[right] * multipliers[idx]);
      const result = Math.max(leftResult, rightResult);
      cache.set(key, result);
      return result;
  }
  return backtrack(0, nums.length - 1, 0);
};
